### PR TITLE
saving classifications

### PR DESCRIPTION
**Pull request by elena-vi**
Part two of:
https://github.com/LBHackney-IT/lbh-income-api/pull/213

once classifications have been calculated, this needs to be saved into the database

### CODE REVIEW

File: db/migrate/20191008164749_add_classification_to_case_priority.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddClassificationToCasePriority < ActiveRecord::Migration[5.2]
+  def change
+    add_column :case_priorities, :classification, :integer
```
Comment (maysakanoni) [inline db/migrate/20191008164749_add_classification_to_case_priority.rb]: This is more of a question than a comment, do we save symbols as integers in the db?

Comment (elena-vi) [reply db/migrate/20191008164749_add_classification_to_case_priority.rb]: ✨enum magic ✨
https://naturaily.com/blog/ruby-on-rails-enum

File: lib/hackney/income/stored_tenancies_gateway.rb [hunk 2]
```diff
@@ -8,6 +8,7 @@ def store_tenancy(tenancy_ref:, priority_band:, priority_score:, criteria:, weig
           criteria,
           weightings
         )
+        classification_usecase = Hackney::Income::TenancyPrioritiser::TenancyClassification.new(criteria)
```
Comment (richrace) [inline lib/hackney/income/stored_tenancies_gateway.rb]: Style suggestion: Whitespace between line `11-12` 🙃

File: app/models/hackney/income/models/case_priority.rb [hunk 3]
```diff
@@ -2,6 +2,8 @@ module Hackney
   module Income
     module Models
       class CasePriority < ApplicationRecord
+        enum classification: %i[no_action send_letter_two send_letter_one send_first_SMS send_NOSP send_warning_letter]
```
Comment (richrace) [inline app/models/hackney/income/models/case_priority.rb]: As these are used here https://github.com/LBHackney-IT/lbh-income-api/blob/257e89607b77765976a18a86611e7f62533a03f0/lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb#L11-L16 could we look at making them either Constants or using enum values `Income::Models::CasePriority.classifications[:no_action]` ?

This should make it easier if we want to add any in the future.

File: lib/hackney/income/stored_tenancies_gateway.rb [hunk 4]
```diff
@@ -8,6 +8,8 @@ def store_tenancy(tenancy_ref:, priority_band:, priority_score:, criteria:, weig
...
```
Comment (kjdchapman) [inline lib/hackney/income/stored_tenancies_gateway.rb]: I know it hasn't been introduced in this PR, but I don't "get" this use case name quickly. I'm also a little confused by the namespace. 

I would have expected this to be something like `Hackney::Income::UseCase::ClassifyTenancy`.

Comment (kjdchapman) [inline lib/hackney/income/stored_tenancies_gateway.rb]: Relying directly on a Use Case in a Gateway will tie high-level policy to low-level details, which could make this difficult to work with in the future. 

I'm worried about both the components involved (I'd expect a use case to use a gateway, not the other way around) and the lack of dependency injection.

I haven't looked at the entire architecture around this component, so I may be missing something.

File: lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb [hunk 5]
```diff
@@ -7,12 +7,15 @@ def initialize(criteria)
...
         def execute
-          return :send_NOSP if send_nosp?
-          return :send_warning_letter if send_warning_letter?
-          return :send_letter_two if send_letter_two?
-          return :send_letter_one if send_letter_one?
-          return :send_first_SMS if send_sms?
-          :no_action
+          case_classifications = Hackney::Income::Models::CasePriority.classifications
+
+          return case_classifications[:send_NOSP]           if send_nosp?
```
Comment (kjdchapman) [inline lib/hackney/income/tenancy_prioritiser/tenancy_classification.rb]: As discussed irl, this strikes me as the "wrong" control flow to express this logic, but it isn't particularly unclear when read. 

I'd be happy with us releasing this, and I'm happier now it's tied back to the enum of the relevant class.

File: spec/lib/hackney/income/tenancy_prioritiser/tenancy_classification_spec.rb [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
   let(:criteria) { Stubs::StubCriteria.new }
   let(:assign_classification) { described_class.new(criteria) }
+  let(:classifications) { Hackney::Income::Models::CasePriority.classifications }
```
Comment (kjdchapman) [inline spec/lib/hackney/income/tenancy_prioritiser/tenancy_classification_spec.rb]: I'm not sure if we should include this in the tests or not - I'd lean towards no. 

Including it here ties the tests to a particular classes' details, and unlike the production code, I'm not worried about getting the symbols wrong (as my tests would then fail).

Comment (richrace) [reply spec/lib/hackney/income/tenancy_prioritiser/tenancy_classification_spec.rb]: So using  `Hackney::Income::Models::CasePriority.classifications[:no_action]` will return the enum value (`0` in this case) so the tests would read something like `eq(0)` or `eq(1)` which are not readable.

File: lib/hackney/income/stored_tenancies_gateway.rb (hunk 4 above)
Comment (richrace) [reply lib/hackney/income/stored_tenancies_gateway.rb]: I guess this is the same problem as Line 7?

File: spec/lib/hackney/income/tenancy_prioritiser/tenancy_classification_spec.rb (hunk 6 above)
Comment (kjdchapman) [reply spec/lib/hackney/income/tenancy_prioritiser/tenancy_classification_spec.rb]: Ah, fair enough. Is it a design smell that the result of `use_case.execute` is a meaningless number?

Comment (richrace) [reply spec/lib/hackney/income/tenancy_prioritiser/tenancy_classification_spec.rb]: Having a conversation with @kjdchapman we have agreed that the Use Case's return value not being readable is more of an issue that using the defined enum list.

Also, we should have a spec covering the case classification in `spec/lib/hackney/income/stored_tenancies_gateway_spec.rb`

File: lib/hackney/income/stored_tenancies_gateway.rb (hunk 4 above)
Comment (richrace) [reply lib/hackney/income/stored_tenancies_gateway.rb]: Discussed with @kjdchapman and we're going to leave this alone for now and it's probably quite a bit to tackle which is not the scope of this PR

**Pull request merged**